### PR TITLE
Use PaperLib to load chunks and teleport players async

### PR DESCRIPTION
**Pull request by DoNotSpamPls**
squash merge please - i messed up my repository and stuff
Resolves-ish #2326 and #2193

### CODE REVIEW

File: Essentials/src/com/earth2me/essentials/utils/LocationUtil.java [hunk 1]
```diff
@@ -103,11 +96,12 @@ public static boolean isBlockUnsafeForUser(final IUser user, final World world,
         return isBlockAboveAir(world, x, y, z);
     }
 
-    public static boolean isBlockUnsafe(final World world, final int x, final int y, final int z) {
+    private static boolean isBlockUnsafe(final World world, final int x, final int y, final int z) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/utils/LocationUtil.java]: LocationUtil static methods should be public.

File: Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java [hunk 2]
```diff
@@ -642,7 +643,7 @@ public void run() {
                     while (LocationUtil.isBlockDamaging(loc.getWorld(), loc.getBlockX(), loc.getBlockY() - 1, loc.getBlockZ())) {
                         loc.setY(loc.getY() + 1d);
                     }
-                    user.getBase().teleport(loc, TeleportCause.PLUGIN);
+                    PaperLib.teleportAsync(user.getBase(), loc);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java]: Is there absolutely no way to specify the TeleportCause?

File: Essentials/pom.xml [hunk 3]
```diff
@@ -95,12 +107,6 @@
             <artifactId>ReflectionProvider</artifactId>
             <version>2.16.0</version>
             <scope>compile</scope>
-            <exclusions>
-                <exclusion>
-                    <groupId>org.bukkit</groupId>
-                    <artifactId>craftbukkit</artifactId>
-                </exclusion>
-            </exclusions>
```
Comment (mdcfe) [inline Essentials/pom.xml:103]: What actual effect does this have?

File: EssentialsSpawn/pom.xml [hunk 4]
```diff
@@ -13,13 +13,52 @@
...
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <createDependencyReducedPom>false</createDependencyReducedPom>
+                    <relocations>
+                        <relocation>
+                            <pattern>io.papermc.lib</pattern>
```
Comment (mdcfe) [inline EssentialsSpawn/pom.xml]: We can use the shaded PaperLib from the Essentials module - no need to shade it twice.

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: Essentials/pom.xml (hunk 3 above)
Comment (DoNotSpamPls) [reply Essentials/pom.xml:103]: That module doesn't have craftbukkit as a dependency, so excluding CB here does absolutely nothing - that's why I have removed it (and the bukkit dependency there is omitted automatically for conflicting with the 1.13 dependency in the main project)

File: Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java (hunk 2 above)
Comment (DoNotSpamPls) [reply Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java]: Currently not. I could ask the Paper devs to add such method. What is the cause used for though?

Comment (DoNotSpamPls) [reply Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java]: Fixed.

File: Essentials/src/com/earth2me/essentials/utils/LocationUtil.java [hunk 5]
```diff
@@ -49,46 +49,39 @@ public static ItemStack convertBlockToItem(final Block block) {
...
+        } catch (NoSuchMethodError ignored) {} // failing now :(
         if (block == null) {
             throw new Exception("Not targeting a block");
         }
         return block.getLocation();
     }
 
-    static boolean isBlockAboveAir(final World world, final int x, final int y, final int z) {
+    public static boolean isBlockAboveAir(final World world, final int x, final int y, final int z) {
+        PaperLib.getChunkAtAsync(world, x, z); // Get the chunk first so it can be loaded async
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/utils/LocationUtil.java]: This call is useless when you simply do getBlockAt in the next line. This function would need to chain off of the getChunkAtAsync  and return CompletableFuture<Boolean>.

File: Essentials/src/com/earth2me/essentials/utils/LocationUtil.java [hunk 6]
```diff
@@ -108,6 +101,7 @@ public static boolean isBlockUnsafe(final World world, final int x, final int y,
     }
 
     public static boolean isBlockDamaging(final World world, final int x, final int y, final int z) {
+        PaperLib.getChunkAtAsync(world, x, z); // Get the chunk first so it can be loaded async
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/utils/LocationUtil.java]: Same as line 84.

**Review by Ichbinjoe — CHANGES_REQUESTED** [review]

File: Essentials/src/com/earth2me/essentials/utils/LocationUtil.java (hunk 5 above)
Comment (Ichbinjoe) [reply Essentials/src/com/earth2me/essentials/utils/LocationUtil.java]: If the chunk isn't loaded, getChunkAtAsync should queue up the chunk load, however the next line then immediately _has_ to load the chunk in sync.

**Comment by DoNotSpamPls** [comment]
Sadly, I can't use the async chunk load methods without potentially rewriting every single use of the utility methods, which is not something I'm looking forward to do at the moment. Maybe some other day 🤔

**Review by Ichbinjoe — APPROVED** [review]

**Review by mdcfe — APPROVED** [review]

**Comment by gp-Airee** [comment]
Why was such a half baked PR merged? This does nothing.

**Comment by DoNotSpamPls** [comment]
@gp-Airee We're well aware the PR doesn't do much. However, to make it work, Essentials needs to be refactored to work with async methods. Work is being done over on [this branch](https://github.com/EssentialsX/Essentials/tree/refactor/async-location-teleport) - feel free to contribute!

**Pull request merged**